tests/fs/api.private.test.ts

```diff
@@ -0,0 +1,284 @@
+          await fs.rm(dirpath)
+
+          expect(await fs.exists(dirpath)).toEqual(false)
+        })
+    )
+  })
+
+  it('writes files to a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('private', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
+          const filepath = path.file('private', 'testDir', data[0])
+          const [val] = data[1]
+
+          await fs.write(filepath, val)
+
+          expect(await fs.exists(filepath)).toEqual(true)
+        })
+    )
+  })
+
+  it('lists files written to a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('private', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
+          const filepath = path.file('private', 'testDir', data[0])
+          const [val] = data[1]
+
+          await fs.write(filepath, val)
+          const listing = await fs.ls(dirpath)
+
+          expect(data[0] in listing).toEqual(true)
+        })
+    )
+  })
+
+  it('moves files into a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('private', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegmentPair(), fileContent()), async data => {
+          const [from, to] = data[0]
+          const [val] = data[1]
+          const fromPath = path.file('private', from)
+          const toPath = path.file('private', 'testDir', to)
+
+          await fs.write(fromPath, val)
+          await fs.mv(fromPath, toPath)
+          const fromExists = await fs.exists(fromPath)
+          const toExists = await fs.exists(toPath)
+
+          expect(toExists && !fromExists).toEqual(true)
+        })
+    )
+  })
+})
+
+
+/* Filesystem */
+
+const emptyFilesystem = async () => {
+  rootKey = await crypto.aes.genKeyStr()
+  return FileSystem.empty({
+    localOnly: true,
+    permissions: {
+      fs: {
+        public: [path.root()],
+        private: [path.root()]
+      }
+    },
+    rootKey
+  })
+}
+
+/* Paths */
+
+const pathSegment = () => {
+  return fc.hexaString({ minLength: 1, maxLength: 20 })
+}
+
+const pathSegmentPair = () => {
+  return fc.set(
+    fc.hexaString({ minLength: 4, maxLength: 20 }),
+    { minLength: 2, maxLength: 2 }
+  )
+}
+
+/* File content */
+
+const fileContent = () => {
+  return fc.frequency(
+    { arbitrary: simpleContent(), weight: 10 },
+    { arbitrary: rawFileContent(), weight: 1 },
+    { arbitrary: recordContent(), weight: 1 }
+  )
+}
+
+const simpleContent = () => {
+  return fc.frequency(
+    { arbitrary: fc.tuple(fc.json(), fc.constant('string')), weight: 6 },
+    { arbitrary: fc.tuple(fc.string({ minLength: 1 }), fc.constant('string')), weight: 4 },
```

Comment:
Empty strings should definitely be tested, too, I think.

Reply:
Yep, and early iteration of the tests was failing, but likely there was something faulty with the test itself. Changing this to include empty strings.